vsintegration/src/FSharp.Editor/CodeLens/FSharpCodeLensService.fs

```diff
@@ -0,0 +1,374 @@
+
+
+open System
+open System.Windows.Controls
+open System.Windows.Media
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Formatting
+open Microsoft.CodeAnalysis
+open System.Threading
+open Microsoft.FSharp.Compiler.SourceCodeServices
+open System.Windows
+open System.Collections.Generic
+open Microsoft.FSharp.Compiler.Range
+open Microsoft.FSharp.Compiler
+open Microsoft.FSharp.Compiler.Ast
+open Microsoft.CodeAnalysis.Editor.Shared.Extensions
+open Microsoft.CodeAnalysis.Editor.Shared.Utilities
+open Microsoft.CodeAnalysis.Classification
+open Internal.Utilities.StructuredFormat
+open System.Windows.Media.Animation
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+open Microsoft.VisualStudio.Text.Classification
+
+type internal CodeLens(taggedText, computed, fullTypeSignature, uiElement) =
+      member val TaggedText: Async<(ResizeArray<Layout.TaggedText> * QuickInfoNavigation) option> = taggedText
+      member val Computed: bool = computed with get, set
+      member val FullTypeSignature: string = fullTypeSignature 
+      member val UiElement: UIElement = uiElement with get, set
+
+type internal FSharpCodeLensService
+    (
+        workspace: Workspace, 
+        documentId: Lazy<DocumentId>,
+        buffer: ITextBuffer, 
+        checker: FSharpChecker,
+        projectInfoManager: FSharpProjectOptionsManager,
+        classificationFormatMapService: IClassificationFormatMapService,
+        typeMap: Lazy<ClassificationTypeMap>,
+        gotoDefinitionService: FSharpGoToDefinitionService,
+        codeLens : CodeLensDisplayService
+     ) as self =
+
+    let lineLens = codeLens
+
+    let visit pos parseTree = 
+        AstTraversal.Traverse(pos, parseTree, { new AstTraversal.AstVisitorBase<_>() with 
+            member __.VisitExpr(_path, traverseSynExpr, defaultTraverse, expr) =
+                defaultTraverse(expr)
+            
+            override __.VisitInheritSynMemberDefn (_, _, _, _, range) = Some range
+
+            override __.VisitTypeAbbrev( _, range) = Some range
+
+            override __.VisitLetOrUse(binding, range) = Some range
+
+            override __.VisitBinding (fn, binding) =
+                Some binding.RangeOfBindingAndRhs
+        })
+
+    let formatMap = lazy classificationFormatMapService.GetClassificationFormatMap "tooltip"
+
+    let mutable lastResults = Dictionary<string, ITrackingSpan * CodeLens>()
+    let mutable firstTimeChecked = false
+    let mutable bufferChangedCts = new CancellationTokenSource()
+    let uiContext = SynchronizationContext.Current
+
+    let layoutTagToFormatting (layoutTag: LayoutTag) =
+        layoutTag
+        |> RoslynHelpers.roslynTag
+        |> ClassificationTags.GetClassificationTypeName
+        |> typeMap.Value.GetClassificationType
+        |> formatMap.Value.GetTextProperties   
+
+    let createTextBox (lens:CodeLens) =
+        asyncMaybe {
+            Async.RunSynchronously(Async.SwitchToContext uiContext, 500)
```

Comment:
❓ How does this cause the code that follows to run on the UI thread? I was expecting to see something like this:

```
do! Async.SwitchToContext uiContext
```

Reply:
Maybe I made a common mistake here but so far I understood this should still work out. I'll change it back and see whether we get more information.

To understand why I did this, I've noticed that the computation did not execute completely (although it should) and tried to find out where I've got a deadlock.

Reply:
@realvictorprm Yes, @sharwell is correct. It's almost always incorrect to do an Async.RunSynchronously inside an async.